File: frontend/api/query/route.py
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel
import os

from langchain_community.vectorstores import SupabaseVectorStore
from langchain_huggingface import HuggingFaceEmbeddings
from langchain.chains import create_retrieval_chain
from langchain.chains.combine_documents import create_stuff_documents_chain
from langchain_core.prompts import ChatPromptTemplate
from supabase.client import create_client, Client
from langchain_google_genai import ChatGoogleGenerativeAI

# Vercel looks for this 'app' variable
app = FastAPI()

# --- Initialize everything ---
supabase_url = os.getenv("SUPABASE_URL")
supabase_key = os.getenv("SUPABASE_SERVICE_KEY")
supabase_client: Client = create_client(supabase_url, supabase_key)
llm = ChatGoogleGenerativeAI(model="gemini-2.5-flash", google_api_key=os.getenv("GOOGLE_API_KEY"), temperature=0)


print("Query API: Loading local embedding model...")
embeddings = HuggingFaceEmbeddings(model_name="all-MiniLM-L6-v2")
print("Query API: Local embedding model loaded.")

vector_store = SupabaseVectorStore(
    client=supabase_client,
    embedding=embeddings,
    table_name="documents",
    query_name="match_documents"
)
# ---

class QueryRequest(BaseModel):
    file_path: str
    query: str
    user_email: str

# This function will be served at POST /api/query
@app.post("/")
async def query_document(request: QueryRequest):
    try:
        file_name = os.path.basename(request.file_path)
        print(f"Querying for user '{request.user_email}' and file '{file_name}'")

        retriever = vector_store.as_retriever(
            search_kwargs={'filter': {
                'user_id': request.user_email,
                'file_name': file_name
            }}
        )

        prompt = ChatPromptTemplate.from_template(
            "Answer the user's question based only on the following context:\n\n<context>{context}</context>\n\nQuestion: {input}"
        )
        document_chain = create_stuff_documents_chain(llm, prompt)
        retrieval_chain = create_retrieval_chain(retriever, document_chain)
        
        response = await retrieval_chain.ainvoke({"input": request.query})
        
        answer = response.get("answer", "I couldn't find an answer in the document.")
        if "I couldn't find an answer" in answer:
            docs = await retriever.ainvoke(request.query)
            if not docs:
                return {"answer": f"Error: No document found with the name '{file_name}' for this user."}
        
        return {"answer": answer}
    except Exception as e:
        print(f"Error during RAG query: {e}")
        raise HTTPException(status_code=500, detail=f"An unexpected error occurred during RAG query: {str(e)}")

File: frontend/app/api/query/route.ts
import { NextResponse } from "next/server";
import { createClient } from "@supabase/supabase-js";
import { HuggingFaceInferenceEmbeddings } from "@langchain/community/embeddings/hf";
import { SupabaseVectorStore } from "@langchain/community/vectorstores/supabase";
import { ChatGoogleGenerativeAI } from "@langchain/google-genai";

const supabase = createClient(
  process.env.NEXT_PUBLIC_SUPABASE_URL!,
  process.env.SUPABASE_SERVICE_ROLE_KEY!
);

const embeddings = new HuggingFaceInferenceEmbeddings({
  apiKey: process.env.HF_API_KEY!,
  model: "sentence-transformers/all-MiniLM-L6-v2",
});

const llm = new ChatGoogleGenerativeAI({
  model: "gemini-2.5-flash",
});

export async function POST(req: Request) {
  try {
    const { query } = await req.json();

    if (!query) {
      return NextResponse.json({ error: "Query is required" }, { status: 400 });
    }

    // Find the most recent document uploaded within the last 120 seconds
    const { data: files, error: fetchError } = await supabase
      .from("documents")
      .select("id, metadata")
      .order("created_at", { ascending: false })
      .limit(1);

    if (fetchError) throw fetchError;
    if (!files || files.length === 0) {
      return NextResponse.json(
        { error: "No recent document found" },
        { status: 404 }
      );
    }

    const recentFile = files[0];
    const uploadedAt = (recentFile.metadata as any)?.uploaded_at;

    if (!uploadedAt) {
      return NextResponse.json(
        { error: "Missing upload timestamp in metadata" },
        { status: 400 }
      );
    }

    const now = new Date();
    const uploadTime = new Date(uploadedAt);
    const diffInSeconds = (now.getTime() - uploadTime.getTime()) / 1000;

    if (diffInSeconds > 120) {
      return NextResponse.json(
        { error: "No recent upload found (older than 120s)" },
        { status: 404 }
      );
    }

    // Load the most recent document’s embeddings
    const vectorStore = await SupabaseVectorStore.fromExistingIndex(embeddings, {
      client: supabase,
      tableName: "documents",
      queryName: "match_documents",
    });

    const results = await vectorStore.similaritySearch(query, 3);

    const context = results.map((r) => r.pageContent).join("\n");

    const response = await llm.invoke(
      `Answer the user's question using the context below:\n\nContext:\n${context}\n\nQuestion: ${query}`
    );

    // Extract text content before returning, with runtime/type guards
    let answer = "No answer generated.";

    if (typeof response === "string") {
      answer = response;
    } else if (
      response &&
      typeof response === "object" &&
      "content" in response &&
      Array.isArray((response as any).content) &&
      (response as any).content.length > 0
    ) {
      const first = (response as any).content[0];
      if (typeof first === "string") {
        answer = first;
      } else if (first && typeof first === "object" && "text" in first) {
        answer = (first as any).text;
      } else {
        answer = JSON.stringify(first);
      }
    } else if (response && typeof response === "object" && "text" in (response as any)) {
      // handle case where response itself has a text field
      answer = (response as any).text;
    }

    return NextResponse.json({ answer });
  } catch (error: any) {
    console.error("❌ Query error:", error);
    return NextResponse.json({ error: error.message }, { status: 500 });
  }
}
